modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/exec/exp/agg/GroupKey.java

```diff
@@ -17,86 +17,98 @@
+    public GroupKey(Row row, RowHandler<Row> hnd) {
+        this.row = row;
+        this.hnd = hnd;
     }
 
     /** */
-    public Object[] fields() {
-        return fields;
+    public Row row() {
+        return row;
+    }
+
+    /** */
+    public RowHandler<Row> rowHandler() {
+        return hnd;
     }
 
     /** {@inheritDoc} */
-    @Override public boolean equals(Object o) {
-        if (this == o)
-            return true;
-        if (o == null || getClass() != o.getClass())
-            return false;
+    @Override public void writeBinary(BinaryWriter writer) throws BinaryObjectException {
+        BinaryRawWriter rawWriter = writer.rawWriter();
 
-        GroupKey grpKey = (GroupKey)o;
+        int colCnt = hnd.columnCount(row);
 
-        return Arrays.equals(fields, grpKey.fields);
+        rawWriter.writeInt(colCnt);
+
+        for (int i = 0; i < colCnt; i++)
+            rawWriter.writeObject(hnd.get(i, row));
     }
 
     /** {@inheritDoc} */
-    @Override public int hashCode() {
-        return Arrays.hashCode(fields);
-    }
+    @Override public void readBinary(BinaryReader reader) throws BinaryObjectException {
+        BinaryRawReader rawReader = reader.rawReader();
 
-    /** */
-    public static Builder builder(int rowLen) {
-        return new Builder(rowLen);
-    }
+        int colCnt = rawReader.readInt();
 
-    /** */
-    public static class Builder {
-        /** */
-        private final Object[] fields;
+        Object[] row0 = new Object[colCnt];
 
-        /** */
-        private int idx;
+        for (int i = 0; i < colCnt; i++)
+            row0[i] = rawReader.readObject();
 
-        /** */
-        private Builder(int rowLen) {
-            fields = new Object[rowLen];
-        }
+        row = (Row)row0;
+        hnd = (RowHandler<Row>)ArrayRowHandler.INSTANCE;
+    }
 
-        /** */
-        public Builder add(Object val) {
-            if (idx == fields.length)
-                throw new IndexOutOfBoundsException();
+    /** {@inheritDoc} */
+    @Override public boolean equals(Object o) {
+        if (this == o)
+            return true;
+        if (o == null || getClass() != o.getClass())
+            return false;
 
-            fields[idx++] = val;
+        GroupKey<Row> other = (GroupKey<Row>)o;
 
-            return this;
-        }
+        int colCnt = hnd.columnCount(row);
 
-        /** */
-        public GroupKey build() {
-            assert idx == fields.length;
+        if (colCnt != other.hnd.columnCount(other.row))
+            return false;
 
-            return new GroupKey(fields);
+        for (int i = 0; i < colCnt; i++) {
+            if (!Objects.equals(hnd.get(i, row), other.hnd.get(i, other.row)))
+                return false;
         }
 
-        /** */
-        public void clear() {
-            Arrays.fill(fields, null);
+        return true;
+    }
 
-            idx = 0;
-        }
+    /** {@inheritDoc} */
+    @Override public int hashCode() {
```

Comment:
Can be cached? Or row may change?

Reply:
Row is immutable. Caching hash code will not bring any benefits. This field is calculated only once - when we put object to hashmap (this statement also checked by JMH - no performance benefits, only additional memory usage)